src/TodoContext.tsx

```diff
@@ -0,0 +1,169 @@
+  setProcessingTodoIds: (loadingTodoIds: number[]) => void,
+  errorMessage: string;
+  setErrorMessage: (errorMessage: string) => void;
+  deleteTodo: (id: number) => void;
+  completedTodos: Todo[];
+  clearAllCompleted: () => void;
+  uncompletedTodosLength: number;
+  handleTodoUpdate: (todo: Todo, todoTitle: string) => void;
+  handleStatusChange: (todo: Todo, status: boolean) => void;
+  setStatusForAll: () => void;
+}
+
+const todoContext: TodoContextType = {
+  todoItems: [],
+  setTodoItems: () => {},
+  tempTodo: null,
+  setTempTodo: () => {},
+  processingTodoIds: [],
+  setProcessingTodoIds: () => {},
+  errorMessage: '',
+  setErrorMessage: () => {},
+  deleteTodo: () => {},
+  completedTodos: [],
+  clearAllCompleted: () => {},
+  uncompletedTodosLength: 0,
+  handleTodoUpdate: () => {},
+  handleStatusChange: () => {},
+  setStatusForAll: () => {},
+};
+
+export const TodoContext = React.createContext(todoContext);
+
+type Props = {
+  children: React.ReactNode;
+};
+
+export const TodoProvider: React.FC<Props> = ({ children }) => {
+  const [todoItems, setTodoItems] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [processingTodoIds, setProcessingTodoIds] = useState<number[]>([]);
+
+  const deleteTodo = (todoId: number) => {
+    setProcessingTodoIds(prevTodoIds => [...prevTodoIds, todoId]);
+    todoService.deleteTodo(todoId)
+      .then(() => setTodoItems(currentTodos => currentTodos
+        .filter(todo => todo.id !== todoId)))
+      .catch((error) => {
+        setErrorMessage(DELETE_ERROR);
+        throw error;
+      })
+      .finally(() => {
+        setProcessingTodoIds(currentTodoIds => currentTodoIds
+          .filter(id => id !== todoId));
+      });
+  };
+
+  const completedTodos = pickCompletedTodos(todoItems);
+
+  const clearAllCompleted = () => {
+    completedTodos.forEach(({ id }) => deleteTodo(id));
+  };
+
+  const uncompletedTodosLength = todoItems.length
+  - completedTodos.length;
+
+  const handleStatusChange = async (todo: Todo, status: boolean) => {
+    const {
+      id,
+      title,
+      userId,
+    } = todo;
+
+    setProcessingTodoIds(prevTodoIds => [...prevTodoIds, id]);
+
+    try {
+      const updatedTodo = await todoService.updateTodo({
+        id,
+        title,
+        userId,
+        completed: status,
+      });
+
+      setTodoItems(prevTodos => prevTodos.map(prevTodo => (
+        prevTodo.id !== updatedTodo.id
+          ? prevTodo
+          : updatedTodo
+      )));
+    } catch (error) {
+      setErrorMessage(UPDATE_ERROR);
+      throw error;
+    } finally {
+      setProcessingTodoIds(currentTodoIds => currentTodoIds
+        .filter(currentTodoId => currentTodoId !== id));
+    }
+  };
+
+  const setStatusForAll = () => {
+    const commonStatus = !!uncompletedTodosLength;
+
+    todoItems.forEach((todo) => handleStatusChange(todo, commonStatus));
```

Comment:
When you need to send a batch of requests it's better to create an array of requests and handle them with `Promise.all` or `Promise.allsettled`